Add IGN_UTILS_VENDOR_CLI11 CMake option to optionally use external CLI11

traversaro opened the pull request:
# 🎉 New feature

This PR adds a CMake option IGN_UTILS_VENDOR_CLI11 (ON by default) that if turned OFF permits to stop vendoring CLI11, and instead rely on a CLI11 already available in the system. 

## Summary
This feature is expected to be used in conda-forge packaging of ign-utils, as CLI11 is already available in conda-forge, and while this library is expected to be used only in executable in ignition projects (see https://github.com/ignitionrobotics/ign-utils/pull/2#issuecomment-758698449), we do not know how users will use it, so just to be safe on the conda-forge side it is better to avoid vendoring CLI11 from the first build.

## Test it
To test it on a conda environment, just create a conda environment and install there cli11:
~~~
conda create -n ignutilsprtest
conda activate ignutilsprtest
conda install cli11 cmake pkg-config compilers libignition-cmake2
git clone https://github.com/traversaro/ign-utils
cd ign-utils
git checkout option_devendor_cli11
mkdir build
cd build
cmake -DIGN_UTILS_VENDOR_CLI11:BOOL=OFF ..
cmake --build . 
cmake --install .
~~~

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

chapulina commented:
Just a high-level comment, I think it would be a bit more organized if the logic to use the external CLI were within the `cli` directory. Would that be possible, or does `ign-cmake` make that complicated?

scpeters commented on CMakeLists.txt:
```diff
@@ -21,19 +21,29 @@ ign_configure_project(VERSION_SUFFIX)
+else()
+  # PKGCONFIG_IGNORE can be removed once pkg-config support in CLI11 (see
+  # https://github.com/CLIUtils/CLI11/pull/523) is released
+  ign_find_package(CLI11 REQUIRED_BY cli PKGCONFIG_IGNORE)
```
I can't tell if the `REQUIRED_BY` argument should be `cli` or `external-cli`

traversaro replied:
This comment should no be solved as `external-cli` is not present anymore in the main CMakeLists.txt .

traversaro commented:
> 
> 
> Just a high-level comment, I think it would be a bit more organized if the logic to use the external CLI were within the `cli` directory. Would that be possible, or does `ign-cmake` make that complicated?

I think it is doable, I just need to shuffle a bit the header locations and play with the `EXCLUDE_FILES` and `EXCLUDE_DIRS` of `ign_install_all_headers` ( https://github.com/ignitionrobotics/ign-cmake/blob/bed3f7b2b2d0210b35808b2cfa17026169f7c4b4/cmake/IgnUtils.cmake#L582 ). This should also help clarify @scpeters comment.

scpeters commented:
> > Just a high-level comment, I think it would be a bit more organized if the logic to use the external CLI were within the `cli` directory. Would that be possible, or does `ign-cmake` make that complicated?
> 
> I think it is doable, I just need to shuffle a bit the header locations and play with the `EXCLUDE_FILES` and `EXCLUDE_DIRS` of `ign_install_all_headers` ( https://github.com/ignitionrobotics/ign-cmake/blob/bed3f7b2b2d0210b35808b2cfa17026169f7c4b4/cmake/IgnUtils.cmake#L582 ). This should also help clarify @scpeters comment.

that sounds good to me

mjcarroll commented:
@traversaro friendly ping

traversaro commented:
> 
> 
> @traversaro friendly ping

Ack, sorry for the delay but I quickly lost track of PRs to which I am not assigned to. Do you have any specific constraint for this PR? I think I can work on it during next week.

mjcarroll commented:
> Do you have any specific constraint for this PR? I think I can work on it during next week.

No rush, going through my backlog :)

traversaro commented:
I resumed working on this, and a side note (not a compliant or anything like that): identify which `CMakeLists.txt` are actually loaded by `` . I tried to follow the code in [`ign_configure_build`](https://github.com/ignitionrobotics/ign-cmake/blob/ignition-cmake2_2.8.0/cmake/IgnConfigureBuild.cmake#L28), but it was not immediately clear, so in the end I just added a few debug prints:
~~~
-- -------> Calling ign_configure_build from main CMakeLists.txt
-- Configuring library: ignition-utils1
-- No tests have been specified for INTEGRATION
-- Adding 1 INTEGRATION tests
-- No tests have been specified for PERFORMANCE
-- No tests have been specified for REGRESSION
-- -------> Executing cli/src/CMakeLists.txt
-- Configuring library: ignition-utils1-cli
-- -------> Executing cli/include/ignition/utils/CMakeLists.txt
-- -------> Running ign_install_all_headers, IGN_UTILS_VENDOR_CLI11: ON
-- The program [cppcheck] was not found! Skipping codecheck setup
-- Build configuration successful
-- -------> Calling ign_create_packages from main CMakeLists.txt
-- Build type: Release
~~~

From that, it seems that the empty file [`cli/include/ignition/utils/cli/CMakeLists.txt`](https://github.com/ignitionrobotics/ign-utils/blob/ignition-utils1_1.0.0/cli/include/ignition/utils/cli/CMakeLists.txt) is not invoked at all? In any case, that's another story, I will probably open an issue for that.

traversaro commented:
I ma

> 
> 
> Just a high-level comment, I think it would be a bit more organized if the logic to use the external CLI were within the `cli` directory. Would that be possible, or does `ign-cmake` make that complicated?

I addressed this request in commit https://github.com/ignitionrobotics/ign-utils/pull/19/commits/d51945f375d15423cbcfa23e422309fdd48e1c89 . Due to the assumptions of `ign-cmake`, this complexified a bit the CMake code, but I hope the final result is satisfiable. 

I also rebased the code and stashed it, so feel free to check the PR as it is to check its difference w.r.t. to the current `ign-utils1` branch, or the commit https://github.com/ignitionrobotics/ign-utils/pull/19/commits/d51945f375d15423cbcfa23e422309fdd48e1c89 to check the changes w.r.t. to the earlier version of the PR.

traversaro commented:
> - [ ] Added tests
> - [ ] Added example and/or tutorial

@mjcarroll this should be ready for review. Not sure if you think that an example and/or tutorial is required for this, as this is something more oriented to ignition-utilis packages, that tipically can read directly the CMakeLists.txt for this kind of options.

> - [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

I reviewed https://github.com/ignitionrobotics/ign-tools/pull/56 that was the first PR of the list, even if that probably does not count. : )

mjcarroll commented:
Thanks! Will take a look.

mjcarroll commented:
This works for me.  To test, I added an example (we can get it in a follow-up PR): https://github.com/ignitionrobotics/ign-utils/commit/f2f6d74ea7ba23d4957160b5ce310b9ac292bb1f

To test un-vendored, I created a colcon workspace with:
repositories:
```
  CLI11:
    type: git
    url: https://github.com/CLIUtils/CLI11
    version: master
  ign-cmake:
    type: git
    url: https://github.com/ignitionrobotics/ign-cmake
    version: ign-cmake2
  ign-utils:
    type: git
    url: https://github.com/ignitionrobotics/ign-utils
    version: cli_example
```

And built with: 
```
         "-DIGN_UTILS_VENDOR_CLI11:BOOL=OFF",
          "-DCLI11_BUILD_TESTS:BOOL=OFF",
          "-DCLI11_BUILD_EXAMPLES:BOOL=OFF"
```

The example was able to build and not use the vendored version.

To test with the vendored version, I changed the `VENDOR_CLI` flag and it worked as expected.
